samples/csharp/getting-started/Ranking_PersonalizedSort/README.md

```diff
@@ -0,0 +1,215 @@
+
+This introductory sample shows how to use ML.NET to predict the relevance and order of hotel search results. In the world of machine learning, this type of prediction is known as ranking.
+
+## Problem
+The ability to perform ranking is a common problem faced by search engines since users expect query results to be ranked\sorted according to their relevance.  This problem extends beyond the needs of search engines to include a variety of business scenarios where personalized sorting is key to the user experience.  Here are a few specific examples:
+* Travel Agency - Provide a list of hotels with those that are most likely to be purchased\booked by the user positioned highest in the list.
+* Shopping - Display items from a product catalog in an order that aligns with a user's shopping preferences.
+* Recruiting - Retrieve job applications ranked according to the candidates that are most qualified for a new job opening.
+
+Ranking is useful to any scenario where it is important to list items in an order that increases the likelihood of a click, purchase, reservation, etc.
+ 
+In this sample, we show how to apply ranking to the first example listed above to rank hotel search results according to the likelihood that the hotel will be purchased\booked by the user. To perform ranking, there are two algorithms currently available - FastTree Boosting (FastRank) and Light Gradient Boosting (LightGbm).  We use the LightGbm's Lambdarank algorithm in this sample to automatically build an ML model to predict ranking. 
+
+## Dataset
+The training and testing data used by this sample is based on a public [dataset available at Kaggle](https://www.kaggle.com/c/expedia-personalized-sort) originally provided by Expedia (https://www.expedia.com).
+
+Expedia's datasets consist of hotel search results that are grouped according to a user's query; each hotel result includes the following details:
+* Hotel attributes, such as location attractiveness and price.
+* User's criteria for searching hotels, such as the number of rooms\children\adults, length of stay, etc.
+* User's purchase and browsing history, such as whether they clicked the link of a hotel or purchased\booked it.
+* Information on similar competitor hotel offerings.
+
+## ML Task - Ranking
+As previously mentioned, this sample uses the LightGbm Lambdarank algorithm which is applied using a supervised learning technique known as "Learning to Rank".  This technique requires that train/test datasets contain groups of data instances that are labeled with their ideal ranking value.  The label is a numerical\ordinal value, such as {4, 3, 2, 1, 0} or a text value {"Perfect", "Excellent", "Good", "Fair", or "Bad"}.  The process for labeling these data instances with their ideal ranking value can be done manually by subject matter experts.  Or, the labels can be determined using other metrics, such as the number of clicks on a given search result. This sample uses the latter approach.
```

Comment:
```suggestion
As previously mentioned, this sample uses the LightGBM LambdaRank algorithm which is applied using a supervised learning technique known as "[Learning to Rank](https://en.wikipedia.org/wiki/Learning_to_rank)".  This technique requires that train/test datasets contain groups of data instances that are each labeled with their relevance scores.  The label is a numerical\ordinal value, such as {4, 3, 2, 1, 0} or a text value {"Perfect", "Excellent", "Good", "Fair", or "Bad"}.  The process for labeling these data instances with their ideal ranking value can be done manually by subject matter experts.  Or, the labels can be determined using other metrics, such as the number of clicks on a given search result. This sample uses the latter approach.
```

Reply:
We may also want to note that it's expected to have many more "Bad" relevance scores than "Perfect". This helps users to avoid converting a ranked list directly into a equally sized bins of {0, 1, 2, 3, 4}.

Reply:
I replaced all references to "ideal ranking value" to "relevance score".

For your point on "This helps to avoid converting a ranked list directly into equally sized bins of {0, 1, 2, 3, 4}" - can you explain a bit more why this needs to be avoided?  I'd like to understand this point and then also explain it in the sample.  For example, would the model potentially end up predicting ranks that are nearly the same for each result?

Good point about reusing scores - I inadvertently left this out and will add it.

Reply:
It is a form of leakage --  I think I'd group the various styles of leakage in the following way...

Leakage styles:
* **Feature availability** (columns which aren't available at prediction time -- e.g. a MonthySalary columns when predicting YearlySalary; or MinutesLate when predicting IsLate; or more subtly NumOfLatePayments when predicting ShouldGiveLoan)
* Non-iid data:
  * **Data leakage** (e.g. splitting a time-series dataset randomly instead of newer data in test set; or MinMax normalizing a dataset then splitting)
  * **Duplicate rows** between train/validation/test (e.g. oversampling a dataset to pad its size before splitting; e.g. different rotations of an single image; bootstrap sampling before splitting; or duplicating rows to up sample minority class) 
  * **Group leakage** -- not including a grouping split column (e.g. Andrew Ng's team had 100k x-rays of 30k patients, meaning ~3 images per patient. They did random splitting instead of ensuring that all images of a patient was in the same split. Hence the model partially memorized the patients instead of learning to recognize pneumonia in chest x-rays. Revised paper had a noticeable drop in scores.)

The first level is if it's column based, or row based leakage. These leakages cause you to misestimate how well your model will perform in production.

This style of leakage is a style of row based leakage, where the lack of a `samplingKeyColumnName` column causes group leakage which may also show up as data leakage or duplicate rows. I'm uncertain which I'd call it; or if my ontology of leakage styles is sufficient.

Reply:
#resolved